shared-libs/couch-request/src/couch-request.js

```diff
@@ -0,0 +1,114 @@
+const request = require('request-promise-native');
+const isPlainObject = require('lodash/isPlainObject');
+const servername = new URL(process.env.COUCH_URL).hostname;
+
+
+const isString = value => typeof value === 'string' || value instanceof String;
+const isTrue = value => isString(value) ? value.toLowerCase() === 'true' : value === true;
+
+const addServername = isTrue(process.env.ADD_SERVERNAME_TO_HTTP_AGENT);
+const methods = {
+  GET: 'GET',
+  POST: 'POST',
+  DELETE: 'DELETE',
+  PUT: 'PUT',
+  HEAD: 'HEAD'
+};
+
+
+const mergeOptions = (target, source, exclusions = []) => {
+  for (const [key, value] of Object.entries(source)) {
+    if (Array.isArray(exclusions) && exclusions.includes(key)) {
+      return target;
+    }
+    target[key] = value; // locally, mutation is preferable to spreading as it doesn't
+    // make new objects in memory. Assuming this is a hot path.
+  }
+  return target;
+};
+
+// When proxying to HTTPS from HTTP (for example where an ingress does TLS termination in an SNI environment),
+// not including a 'servername' for a request to the HTTPS server (eg, def.org) produces the 
+// following error:
+// 
+// 'ERR_TLS_CERT_ALTNAME_INVALID'
+// "RequestError: Error [ERR_TLS_CERT_ALTNAME_INVALID]: Hostname/IP does not match certificate's altnames:
+//  Host: abc.com. is not in the cert's altnames: DNS:def.org"
+// 
+// The addition of 'servername' resolves this error. See docs for 'tls.connect(options[, callback])'
+//  (https://nodejs.org/api/tls.html): "Server name for the SNI (Server Name Indication) TLS extension  It is the
+//  name of the host being connected to, and must be a host name, and not an IP address.".
+// 
+
+const testReqInputs = (method, first, second = {}) => {
+
+  if (Object.hasOwn(methods, method) === false) {
+    return { shouldReject: true, errorMessage: `Unsupported method (${method}) passed to call.` };
+  }
+
+  if (isPlainObject(second) === false) {
+    return { shouldReject: true, errorMessage: '"options" must be a plain object' };
+  }
+
+  const firstIsString = typeof first === 'string' || first instanceof String;
+  
+  if (firstIsString === false && isPlainObject(first) === false) {
+    return { shouldReject: true, errorMessage: '"options" must be a plain object', firstIsString };
+  }
+  
+  return { shouldReject: false, errorMessage: '', firstIsString };
+};
+
+
+const req = (method, first, second = {}) => {
+
+  const { shouldReject, errorMessage, firstIsString = false } = testReqInputs(method, first, second);
```

Comment:
One of the reasons this function is complicated is you're doing two things - one for working out if the first param is a string, and one to validate inputs. This would be cleaner if you cleaned up the parameters (and renamed them from `first`, and `second` before calling the `testReqInputs` function.

Then you can rename `testReqInputs` to `validate` which is actually what you're doing.

Finally you can have validate throw the exception directly which helps get a more accurate stack trace as well as cleaning up the code a lot.

I haven't tested this and it has wider implications but I think you can see what I mean below...

```suggestion
const req = (method, first, options = {}) => {
  if (typeof first === 'string' || first instanceof String) {
    options.url = first;
  }
  try {
    validate(method, options);
  } catch(e) {
    return Promise.reject(e);
  }
```

Reply:
I'll look at the code, but I don't see a better rename for the inputs. 

`first` can either be the url or the `options` object, and I don't have a unifying name for that. `second` can be the `options` object or nothing, but if first is the `options`, then it doesn't make sense for `second` to be named `options`. 

would be nice to have some decent overloading in js. 

I also suspect that can't add another logic branch to the function without triggering sonarqube, but will test that if that's true if the changes make sense.